### ISSUE TITLE
Optimizer stability and epsilon

### ISSUE THREAD

**Issue by jyegerlehner** [ibab/tensorflow-wavenet#149]
Just a heads-up that I might have oversold the Adam and RMSProp optimizer "fixes" in PR128 and PR147, described in issue [143](https://github.com/ibab/tensorflow-wavenet/issues/143).   I made it sound like it completely fixed it. In reality there is a continuum, and a tradeoff to be endured. I still observe jumps in the error. They are less prevalent and smaller in magnitude than before. But still there to some extent.

nakosung pointed out that the tensorflow comments say:

>    The default value of 1e-8 for epsilon might not be a good default in
>    general. For example, when training an Inception network on ImageNet a
>    current good choice is 1.0 or 0.1.

I thought I was radical in cranking it up from default 1e-10 to 1e-5, and 1e-8 to 1e-4 (quite a few orders of magnitude), but apparently there are situations where it needs to be still larger.

I think the tradeoff is that the bigger you make epsilon (and the denominator), the smaller the weight updates are and thus slower the training progress will be. Most times you want the denominator to be able to get small.  I don't have any way of saying that the magic numbers I chose are the exactly right ones.  So we might need to revisit those at some point, or even make the epsilons YACLCO (Yet-Another Command-Line-Configurable Option). 

And finally, I was [able to hack dtype=float16](https://github.com/jyegerlehner/tensorflow-wavenet/tree/fp16) (which, by the way, would be a factor of 2 memory savings) in and get it to run and mostly work, but after training long-enough it became unstable and failed with NaNs. I wonder if bumping up the current epsilons might "fix" that problem. Perhaps that's another reason why we might wish to make the epsilons configurable. I also observed that training wall-clock time was dramatically slower (per step) than with float32, on both Maxwell and Pascal Titan-X. Making me thing perhaps they crippled them for float16 unless you buy GP100 (I have a suspicious mind), so looking further into this ATM is not at the top of my list.

**Comment by wuaalb** [comment]
Interesting, I don't remember ever seeing Adam with `epsilon` higher than 1e-4 before; I wonder how common values such as 0.1 or 1.0 are.

In this thread https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/1JDPbuHN39w one of the authors of the inception v3 paper mentions the unusually high epsilon value of 1.0 was essential for training with *async SGD*, but it's not clear if the same hyper-parameters translate to *sync SGD*. I guess in most cases finding other means to stabilize training would be preferable.

As a side note, another hyper-parameter that's sometimes tuned for Adam is `beta1`, typically 0.9 or 0.5 (or stepped schedule). As far as I understood, higher values reduce variance of the stochastic gradient (inversely proportional to minibatch size) and thus allows reaching a better optimum. However this also introduces a bias which may result in update steps being excessively wrong. The DCGAN paper briefly discusses this for instance.

**Comment by ibab** [comment]
@jyegerlehner: Have you tried one of the previous fixes like swapping out the final `ReLU`s with `TanH`s when testing on `float16`?

**Comment by jyegerlehner** [comment]
@wuaalb Thanks for that info. @nakosung please see wuaalb's comments on epsilon when doing async sgd. I think you were doing some of that.

@ibab, no haven't tried that yet.
